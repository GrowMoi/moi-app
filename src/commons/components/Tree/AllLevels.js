import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, Image, TouchableWithoutFeedback } from 'react-native';
import styled from 'styled-components/native';
import { connect } from 'react-redux';
import { getHeightAspectRatio } from '../../utils';
import NeuronsLayer from './NeuronsLayer';
import neuronActions from '../../../actions/neuronActions';
import Branches from './allBranches';

const CURRENT_TREE_WIDTH = 320;

const Container = styled(TouchableWithoutFeedback)`
`;

const LevelContainer = styled(View)`
  width: 100%;
  height: ${props => getHeightAspectRatio(treeBaseWidth, treeBaseHeight, props.width)};
  position: absolute;
  bottom: 50;
`;

const treeBaseWidth = 1417;
const treeBaseHeight = 1193;
const TreeBase = styled(Image)`
  position: absolute;
  width: ${props => props.width};
  height: ${props => getHeightAspectRatio(treeBaseWidth, treeBaseHeight, props.width)};
  z-index: ${props => props.zIndex || 0};
  overflow: visible;
  justify-content: center;
  align-items: center;
`;

const Levels = styled(View)`
  align-items: center;
  width: 100%;
  height: ${props => getHeightAspectRatio(treeBaseWidth, treeBaseHeight, props.width)};
  overflow: visible;
`;

@connect(state => ({
  label: state.neuron.currentlyPressed,
  device: state.device,
}), {
  setNeuronLabelInfo: neuronActions.setNeuronLabelInfo,
})
export default class AllLevels extends Component {

  componentWillMount() {
    const { setHeightTreeContainer  } = this.props;
    const heightTree = getHeightAspectRatio(treeBaseWidth, treeBaseHeight, CURRENT_TREE_WIDTH);
    setHeightTreeContainer(heightTree);
  }

  shouldComponentUpdate(nextProps, nextState) {
    return nextProps.userTree !== this.props.userTree;
  }

  componentWillReceiveProps(nextProps) {
    if(nextProps.device.dimensions.orientation !== this.props.device.dimensions.orientation) {
      this.props.setNeuronLabelInfo({});
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return nextProps.userTree !== this.props.userTree;
  }

  hideWoodLabel = () => {
    const { setNeuronLabelInfo } = this.props;
    setNeuronLabelInfo({});
  }

  render() {
    const {
      userTree: { tree, meta: { depth } },
    } = this.props;



    const isLevel4 = depth === 4;
    const isLevel3 = depth === 3;
    const isLevel5 = depth === 5;
    const isLevel6 = depth === 6;
    const isLevel7 = depth === 7;
    const isLevel8 = depth === 8;
    const isLevel9 = depth === 9;

    const defaultProps = {
      resizeMode: 'contain',
      width: CURRENT_TREE_WIDTH,
    };

    const treeDimensions = {
      width: treeBaseWidth,
      height: treeBaseHeight,
    };

    return (
      <Container onPress={this.hideWoodLabel}>
      <LevelContainer onPress={this.hideWoodLabel} width={CURRENT_TREE_WIDTH}>
        <Levels width={CURRENT_TREE_WIDTH}>
          {isLevel5 && <TreeBase source={{uri: 'arbol_nivel5_gris'}} {...defaultProps}/>}
          {isLevel4 && <TreeBase source={{uri: 'arbol_nivel4_gris'}} {...defaultProps}/>}
          {isLevel3 && <TreeBase source={{uri: 'arbol_nivel4_gris'}} {...defaultProps}/>}
          {isLevel6 && <TreeBase source={{uri: 'arbol_nivel6_gris'}} {...defaultProps}/>}
          {isLevel7 && <TreeBase source={{uri: 'arbol_nivel7_gris'}} {...defaultProps}/>}
          {isLevel8 && <TreeBase source={{uri: 'arbol_nivel8_gris'}} {...defaultProps}/>}
          {isLevel9 && <TreeBase source={{uri: 'arbol_nivel9_gris'}} {...defaultProps}/>}
          <TreeBase source={{uri: 'arbol_nivel2_color'}} {...defaultProps}/>
          <Branches
            level={depth}
            data={tree}
            treeDimensions={treeDimensions}
            width={CURRENT_TREE_WIDTH}
          />
          <NeuronsLayer
            level={depth}
            data={tree}
            treeDimensions={treeDimensions}
            width={CURRENT_TREE_WIDTH}
          />
        </Levels>
      </LevelContainer>
      </Container>
    );
  }
}

AllLevels.propTypes = {
  userTree: PropTypes.object,
};
