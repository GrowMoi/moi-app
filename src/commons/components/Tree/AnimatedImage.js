import React, { Component } from 'react';
import { Animated, Image } from 'react-native';

export class AnimatedImage extends Component {
  state = {
    animationValue: new Animated.Value(0),
    verticalPosition: 0,
    horizontalPosition: {
      position: 'left',
      flexPosition: 'flex-start'
    }
  }

  animation;

  componentDidMount() {
    const { infiniteLoop, minValue, delayStart = 0 } = this.props;

    if (minValue) this.setState({ animationValue: new Animated.Value(minValue) });

    this.animation = this.createAnimation();
    setTimeout(() => {
      this.initAnimation(infiniteLoop);
    }, delayStart);
  }

  componentWillUnmount() {
    this.animation.stop();
    this.animation = null;
  }

  createAnimation() {
    const { maxValue } = this.props;
    const { animationValue } = this.state;
    return Animated.timing(
      animationValue,
      {
        toValue: maxValue || 300,
        duration: this.getAnimationTime()
      }
    );
  }

  initAnimation(infiniteLoop) {
    const { minValue, verticalPosition,  horizontalPosition } = this.props;
    if(!this.animation) return;
    this.animation.start(() => {
      if (infiniteLoop) {
        this.setState({
          animationValue: new Animated.Value(minValue || 0),
          verticalPosition: verticalPosition(),
          horizontalPosition: horizontalPosition(),
          });
        this.animation = this.createAnimation();
        this.initAnimation(true);
      }
    });
  }

  getAnimationTime() {
    const { timeRange, animationTime = 1000 } = this.props;
    if(timeRange) {
      return Math.random() * (timeRange[1] - timeRange[0]) + timeRange[0];
    } else {
      return animationTime;
    }
  }

  render() {
    const { source } = this.props;
    const { animationValue, verticalPosition, horizontalPosition } = this.state;
    console.log("TCL: AnimatedImage -> render -> source", source)

    return (
      <Animated.View
        style={{
          position: 'absolute',
          top: verticalPosition,
          justifyContent: horizontalPosition.flexPosition,
          alignItems: horizontalPosition.flexPosition,
          [horizontalPosition.position]: animationValue,
        }}
      >
        <Image source={{uri: source}} style={{width: 80, height: 60}} />
      </Animated.View>
    );
  }
}
