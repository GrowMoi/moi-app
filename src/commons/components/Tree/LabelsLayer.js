import React, { Component } from 'react'
import { Text, View } from 'react-native'
import { connect } from 'react-redux';
import { Actions } from 'react-native-router-flux';
import WoodLabel from '../WoodLabel/WoodLabel';
import neuronActions from '../../../actions/neuronActions'
import Preloader from '../Preloader/Preloader';
import deviceUtils from '../../utils/device-utils';
import { ActionConst } from 'react-native-router-flux'

const isTablet = deviceUtils.isTablet();

class LabelsLayer extends Component {

  state = {
    loading: false,
  }

  hideWoodLabel = () => {
    const { setNeuronLabelInfo } = this.props;
    setNeuronLabelInfo({});
  }

  playContent = async () => {
    const { label } = this.props;

    await this.getCurrentContents(label.id);
    Actions.content({ title: label.title, neuron_id: label.id, type: ActionConst.RESET });
  }

  getCurrentContents = async (neuronId) => {
    const { loadNeuronByIdAsync } = this.props;
    this.setState({ loading: true });
    try {
      await loadNeuronByIdAsync(neuronId);
      this.setState({ loading: false });

    } catch (error) {
      this.setState({ loading: false });
      console.log(error.message);
    }
  }

  render() {
    const { label, zoomScale, zoomInfo, tree: { meta: { depth } } } = this.props;
    const { loading } = this.state;

    const neuronWidth = label.width;
    const isBigNeuron = neuronWidth >= (isTablet ? 160 : 80);
    const zoom = zoomScale === 3 ? 4 : zoomScale

    const zoomScaleValues = {
      1: isTablet ? 2.1 : 1.2,
      4: isTablet ? 1.2 + (zoomInfo.scale / 3) : 0.7 + (zoomInfo.scale / 5),
    }

    const topPositions = {
      1: isTablet ? -70 : -62,
      4: isTablet ? -75 : -65,//ok both
    }

    const leftPositions = {
      1: isTablet ? isBigNeuron ? 38 : 15 : isBigNeuron ? 3 : -15,
      4: isTablet ? - 21 : - 30,//ok both
    }

    return (
      <View style={{ flex: 1, position: 'absolute', top: 1, left: 1 }}>
        {label.pageX && label.pageY &&
          <WoodLabel
            text={label.title}
            onPress={() => {
              this.hideWoodLabel();
              this.playContent();
            }}
            style={{
              position: 'absolute',
              top: label.pageY + topPositions[zoom],
              left: label.pageX + leftPositions[zoom],
              transform: [{
                scale: zoomScaleValues[zoom],
              }],
            }}
          />
        }
        {loading && (<Preloader />)}
      </View>
    )
  }
}

const mapStateToProps = (state) => ({
  label: state.neuron.currentlyPressed,
  tree: state.tree.userTree,
  device: state.device,
  zoomInfo: state.tree.zoomInfo,
  zoomScale: state.tree.zoomScale,
})

const mapDispatchToProps = {
  setNeuronLabelInfo: neuronActions.setNeuronLabelInfo,
  loadNeuronByIdAsync: neuronActions.loadNeuronByIdAsync,
}

export default connect(mapStateToProps, mapDispatchToProps)(LabelsLayer)
