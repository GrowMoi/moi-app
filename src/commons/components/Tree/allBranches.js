import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, Image } from 'react-native';
import styled from 'styled-components/native';
import { getHeightAspectRatio } from '../../utils';
import levelsConfig from './neuronConfigs/levels.config';
import { FLORECIDA } from '../../../constants';

const Container = styled(View)`
  position: absolute;
  width: ${({ width }) => width};
  height: ${({ treeDimensions: { width: _width, height }, width }) => getHeightAspectRatio(_width, height, width)};
  bottom: 50;
  left: 0;
  overflow: visible;
`;

const FloweredBranch = styled(Image)`
  width: ${({ width }) => width};
  height: ${({ treeDimensions: { width: _width, height }, width }) => getHeightAspectRatio(_width, height, width)};
  position: absolute;
  bottom: 0;
  left: 0;
`;

const FloweredBranchContainer = styled(View)`
  width: ${({ width }) => width};
  height: ${({ treeDimensions: { width: _width, height }, width }) => getHeightAspectRatio(_width, height, width)};
  position: absolute;
  bottom: 0;
  left: 0;
`;

const RecursiveBranches = ({
  data,
  level = 2,
  maxLevel,
  configs,
  index = 0,
  width,
  treeDimensions,
  parentIndex = 0,
  path = '',
}) => {
  let children;

  const isFlowered = data.state === FLORECIDA;
  // const parentIndexExistInConfigs = `${parentIndex}` in configs[`level${level}`];
  // const indexExistInConfigs = parentIndexExistInConfigs && `${index}` in configs[`level${level}`][parentIndex];

  let newPath;
  if (path) {
    newPath = `${path}.${level}.${index}`;
  } else {
    newPath = `${level}.${index}`;
  }

  if (data.children && data.children.length) {
    children = data.children.map((child, childIndex) => {
      if (level === maxLevel) return null;
      return (
        <RecursiveBranches
          key={`branch-level${level}-${index}-${childIndex}`}
          data={child}
          level={level + 1}
          maxLevel={maxLevel}
          configs={configs}
          parentIndex={index}
          index={childIndex}
          width={width}
          treeDimensions={treeDimensions}
          path={newPath}
        />
      );
    });
  }


  return (
    <FloweredBranchContainer width={width} treeDimensions={treeDimensions}>
      {isFlowered && ((configs[newPath] || {}).floweredBranches || []).map((sourceBranch, branchIndex) => {
        return (
          <FloweredBranch
            key={`branch-level${level}-${parentIndex}-${index}-branch${branchIndex}`}
            source={sourceBranch}
            width={width}
            treeDimensions={treeDimensions}
            resizeMode='contain'
          />
        );
      })}
      {children}
    </FloweredBranchContainer>
  );
};

export default class Branches extends Component {
  renderBranchByDirection = (direction, data) => {
    const levelConfig = levelsConfig[direction];
    const { treeDimensions, width, level } = this.props;

    return (
      <RecursiveBranches
        key={`branch-${direction}`}
        data={data}
        maxLevel={level}
        width={width}
        treeDimensions={treeDimensions}
        configs={levelConfig}
      />
    );
  }

  renderBranches = () => {
    const { data: { root } } = this.props;

    const branches = root.children.map((child, i) => {
      if (i === 0) return this.renderBranchByDirection('left', child, i);
      else if (i === 1) return this.renderBranchByDirection('leftCenter', child, i);
      else if (i === 2) return this.renderBranchByDirection('rightCenter', child, i);
      else if (i === 3) return this.renderBranchByDirection('right', child, i);
    });

    return branches;
  }

  render() {
    const { treeDimensions, width } = this.props;

    return (
      <Container width={width} treeDimensions={treeDimensions}>
        {this.renderBranches()}
      </Container>
    );
  }
}

Branches.propTypes = {
  data: PropTypes.object,
  width: PropTypes.number,
  level: PropTypes.number,
  treeDimensions: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
  }),
};
