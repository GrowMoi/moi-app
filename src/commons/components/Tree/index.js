export default from './Tree';
export { default as Neuron } from './Neuron';
