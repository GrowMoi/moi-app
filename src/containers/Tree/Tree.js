import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';
import { Text, TouchableOpacity } from 'react-native';
// import { ScreenOrientation } from 'expo';
import { Actions } from 'react-native-router-flux';
import Navbar from '../../commons/components/Navbar/Navbar';

const WrapperScene = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
  position: relative;
`;

export default class Tree extends Component {
  componentDidMount() {
    console.log(this.props);
  }

  toggle

  render() {
    return (
      <WrapperScene>
        <Navbar />
        <TouchableOpacity>
          <Text color='white' onPress={() => Actions.refresh({ key: 'moiDrawer', open: value => !value })}>Press Me!</Text>
        </TouchableOpacity>
      </WrapperScene>
    );
  }
}

Tree.propTypes = {
  title: PropTypes.string,
};
