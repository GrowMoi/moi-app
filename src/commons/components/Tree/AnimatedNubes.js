import React, { Component } from 'react';
import { View } from 'react-native';
import { AnimatedImage } from './AnimatedImage';
import uuid from 'uuid/v4';
import { Size } from '../../styles';

const nubesList = ['nube1l', 'nube2l', 'nube3l'];
const delayStartRange = [0, 4000];
const positionNube = [{
  position: 'left',
  flexPosition: 'flex-start'
}, {
  position: 'right',
  flexPosition: 'flex-end'
}];

export class AnimatedNubes extends Component {
  getRandomTopPostion = () => {
    const { deviceHeight, orientation } = this.props;

    const minValue = 0;
    const maxValue = orientation === 'LANDSCAPE' ? 15 : 30;

    const randomPercentage = Math.floor(Math.random() * (maxValue - minValue + 1)) + minValue;
    return (deviceHeight * randomPercentage) / 100;
  }

  getNubesArray() {
    return nubesList.sort((a, b) => {
      return 0.5 - Math.random();
    });
  }

  getDelayStart() {
    return Math.random() * (delayStartRange[1] - delayStartRange[0]) + delayStartRange[0];
  }

  getNubesData() {
    return this.getNubesArray().map(
      nube => {
        return {
          source: nube,
          delayStart: this.getDelayStart(),
          verticalPosition: this.getRandomTopPostion(),
          horizontalPosition: this.getNubePosition(),
        }
      }
    );
  }


  getAnimationValues() {
    return {
      minValue: 0,
      maxValue: this.getWidthContainerNube(),
    }
  }

  getWidthContainerNube() {
    return this.props.deviceWidth + (this.props.deviceWidth * 0.7);
  }

  getNubePosition() {
    return positionNube[Math.floor(Math.random()*positionNube.length)];
  }

  render() {
    const { orientation } = this.props;

    const constainerStyle = {
      position: "relative",
      flex: 1,
      left: -130,
      width: this.getWidthContainerNube(),
      flexDirection: 'column',
      top: Size.mavbarTopSpace,
    };
    const animationValues = this.getAnimationValues();
    const isLandscape = orientation === 'LANDSCAPE';

    return (
      <View style={constainerStyle}>
        {this.getNubesData().map(
          (nube, i) =>
            <AnimatedImage
              key={uuid()}
              source={nube.source}
              minValue={animationValues.minValue}
              maxValue={animationValues.maxValue}
              timeRange={isLandscape ? [25000, 40000] : [15000, 30000]}
              delayStart={nube.delayStart}
              verticalPosition={this.getRandomTopPostion}
              horizontalPosition={this.getNubePosition}
              infiniteLoop
            />
        )}
      </View>
    );
  }
}