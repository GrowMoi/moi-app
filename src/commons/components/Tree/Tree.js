import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, StyleSheet, AsyncStorage, PixelRatio, ImageEditor, ImageStore, Platform } from 'react-native';
import { takeSnapshotAsync } from 'expo';
import styled from 'styled-components/native';
import { connect } from 'react-redux';
import ViewTransformer from 'react-native-view-transformer-next';
import { Actions } from 'react-native-router-flux';
import { isTablet } from 'react-native-device-detection';
import { Maceta } from '../SceneComponents';
import treeActions from '../../../actions/treeActions';
import Preloader from '../Preloader/Preloader';
import { Video } from '../../../commons/components/VideoPlayer';
import vineta_1 from '../../../../assets/videos/vineta_1.mp4';
import LabelsLayer from './LabelsLayer'

// Levels
import Level1 from './Level1';
import Level2 from './Level2';
import Level3 from './Level3';
import AllLevels from './AllLevels';
import userActions from '../../../actions/userActions';
import neuronActions from '../../../actions/neuronActions';
import { AnimatedNubes } from './AnimatedNubes';
import { Sound } from '../SoundPlayer';
import NewAchievementsModal from '../Quiz/NewAchievements';
import EventCompletedModal from '../Quiz/EventCompleted';
import { Size } from '../../styles';

const styles = StyleSheet.create({
  treeView: {
    position: 'relative',
    flex: 1,
  },
});

const TreeContainer = styled(View)`
  flex: 1;
  justify-content: flex-end;
`;

const MacetaContainer = styled(View)`
  position: absolute;
  align-items: center;
  left: 0;
  right: 0;
  bottom: 0;
`;

@connect(store => ({
  device: store.device,
  userTree: store.tree.userTree,
  user: store.user.userData,
  quizResult: store.user.quizResult,
}), {
    loadTreeAsync: treeActions.loadTreeAsync,
    getUserProfileAsync: userActions.getUserProfileAsync,
    getAchievementsAsync: userActions.getAchievementsAsync,
    setZoomTreeInfo: treeActions.setZoomTreeInfo,
    setZoomScale: treeActions.setZoomScaleTree,
    setNeuronLabelInfo: neuronActions.setNeuronLabelInfo,
    uploadTreeImageAsync: userActions.uploadTreeImageAsync,
    removeQuizResult: userActions.removeQuizResult,
  })
export default class Tree extends Component {

  treeView = null;

  state = {
    loading: true,
    hasUserTree: false,
    level: null,
    zoomScale: 1,
    modalVisible: false,
    events: [],
    showModalNewAchievements: false,
    showEventCompleted: false,
    eventTitle: '',
    achievements: null,
    showLabelLayer: true,
  }

  initialActions = async () => {
    const { loadTreeAsync, getUserProfileAsync, getAchievementsAsync, user } = this.props;

    await Promise.all([
      loadTreeAsync(),
      getUserProfileAsync(user.profile.id),
      getAchievementsAsync(),
    ]);
  }

  async componentDidMount() {
    this.setTitleView();
    await this.initialActions();
    this.getTreeLevel();
    this.setState({ loading: false });
    this.handleVideoFirstLogin();
  }

  componentWillReceiveProps(newProps) {
    const { userTree } = this.props;
    const newUserTree = newProps.userTree;

    if (Object.keys(userTree).length !== 0 && userTree !== newUserTree) {
      setTimeout(() => { this.takeScreenShotTree() }, 2000);
    }
  }

  setTitleView() {
    const { user } = this.props;
    Actions.refresh({ title: user.profile.username });
  }

  handleVideoFirstLogin = async () => {
    const { userTree } = this.props;

    if (!userTree.tree) return;

    const videoShown = await AsyncStorage.getItem('videoShown');
    if (!videoShown && userTree.tree.root.learnt_contents === 0) this.showVideo();
  }

  showVideo = (show = true) => {
    this.setState({ modalVisible: show });
  }

  onPlaybackStatusUpdate = async (playbackStatus) => {
    if (playbackStatus.didJustFinish) {
      await AsyncStorage.setItem('videoShown', "true");
      this.showVideo(false);
    }
  }

  prevZoomInfo = null;
  canPlaySound = true;

  onViewTransformed = zoomInfo => {
    const { setZoomTreeInfo, setNeuronLabelInfo } = this.props;
    if (this.prevZoomInfo === null) {
      this.prevZoomInfo = zoomInfo;
    }

    if (this.isSameZoomInfo(this.prevZoomInfo, zoomInfo)) return;
    if (this.canPlaySound) {
      // setNeuronLabelInfo({});
      this.setState({showLabelLayer: false});
      Sound.playOverBackgroundSound('treeActions', true, 1);
      this.canPlaySound = false;
    }
    this.prevZoomInfo = zoomInfo;
  }

  isSameZoomInfo(previousZoomInfo, newZoomInfo) {
    return previousZoomInfo.scale === newZoomInfo.scale && previousZoomInfo.translateX === newZoomInfo.translateX && previousZoomInfo.translateY === newZoomInfo.translateY;
  }

  onTransformGestureReleased = (zoomInfo) => {
    const { setZoomTreeInfo, setNeuronLabelInfo } = this.props;
    setZoomTreeInfo(zoomInfo);
    setTimeout(() => {
      Sound.stopOverBackgroundSound();
      this.canPlaySound = true;
      setNeuronLabelInfo({});
      this.setState({showLabelLayer: true});
    }, 1000);
  }

  selectCurrentLevel = (userTree) => {
    const level = userTree.meta.depth;
    const levelData = { userTree };

    const levels = {
      tree1: <Level1 {...levelData} />,
      tree2: <Level2 {...levelData} />,
      tree3: <Level3 {...levelData} />,
      allLevels: <AllLevels {...levelData} zoomScale={4} />,
    };

    switch (level) {
      case 1:
        return { component: levels.tree1, zoomScale: 1 };
      case 2:
        return { component: levels.tree2, zoomScale: 1 };
      case 3:
        return { component: levels.allLevels, zoomScale: 3 };
      case 4:
        return { component: levels.allLevels, zoomScale: 3 };
      case 5:
        return { component: levels.allLevels, zoomScale: 4 };
      case 6:
        return { component: levels.allLevels, zoomScale: 4 };
      case 7:
        return { component: levels.allLevels, zoomScale: 4 };
      case 8:
        return { component: levels.allLevels, zoomScale: 4 };
      case 9:
        return { component: levels.allLevels, zoomScale: 4 };
      default:
        return { component: levels.tree1, zoomScale: 1 };
    }
  }

  getTreeLevel = () => {
    const { userTree, setZoomScale } = this.props;

    if (userTree.meta) {
      const currentLevel = this.selectCurrentLevel(userTree);

      this.setState({
        hasUserTree: true,
        level: currentLevel.component,
        zoomScale: currentLevel.zoomScale,
      });

      setZoomScale(currentLevel.zoomScale);
    }
  }

  async takeScreenShotTree() {
    const { device: { dimensions: { width, height, orientation } } } = this.props;

    const pixelRatio = PixelRatio.get(); // The pixel ratio of the device

    const resultScreenShot = await takeSnapshotAsync(this.treeView, {
      result: 'base64',
      height: (height / pixelRatio),
      width: width / pixelRatio,
      quality: 1,
      format: 'png',
    });

    // if (orientation === 'LANDSCAPE' || Platform.OS === 'android') {
    this.uploadTreeImage(resultScreenShot);
    return;
    // }

    // this.cropImage(this.normalizeBase64Image(resultScreenShot), width, height);
  }

  getHeigthImage(height) {
    const { userTree } = this.props;

    const level = userTree.meta.depth;
    const imageHeight = height / 1.9617;

    const levelsOffset = {
      1: imageHeight / 2.125,
      2: imageHeight / 8.5,
      3: imageHeight / 2.2667,
      4: imageHeight / 2.2667,
      5: imageHeight / 4.25,
      6: imageHeight / 6.8,
      7: imageHeight / 13.6,
      8: 0,
      9: 0,
    }

    return {
      offset: levelsOffset[level],
      height: levelsOffset[level] === 0 ? height / 2 : imageHeight - levelsOffset[level],
    }
  }

  async cropImage(image, width, height) {

    const heightImage = this.getHeigthImage(height);

    let cropData = {
      offset: { x: 0, y: 0 },
      size: { width: width, height: height }
    }

    ImageEditor.cropImage(image, cropData, this.getImageFromRctImage, (error) => { console.log('error ===> ', error) });
  }

  getImageFromRctImage = (rctImageUri) => {
    ImageStore.getBase64ForTag(rctImageUri, (image) => {
      ImageStore.removeImageForTag(rctImageUri);
      this.uploadTreeImage(image);
    }, () => { });
  }

  uploadTreeImage = async (image) => {
    const { uploadTreeImageAsync, getUserProfileAsync, user } = this.props;
    await uploadTreeImageAsync(this.normalizeBase64Image(image));
    await getUserProfileAsync(user.profile.id);
    console.log("upload ok");
  }

  normalizeBase64Image(base64Image) {
    return 'data:image/png;base64,' + base64Image.replace(/(?:\r\n|\r|\n)/g, '')
  }

  validateResultQuiz() {
    const { quizResult: { achievements = [], event = {} } } = this.props;
    if (achievements.length > 0) {
      this.showModalNewAchievements(achievements);
    }

    if (event.completed) {
      this.showEventCompletedModal(event.info.title);
    }

    if (achievements.length === 0 && !event.completed) {
      this.removeQuizResult();
    }
  }

  showModalNewAchievements = (achievements) => {
    this.setState({ showModalNewAchievements: true, achievements: achievements });
  }

  hideModalNewAchievements = () => {
    if (!this.state.showEventCompleted) {
      this.removeQuizResult();
    }
    this.setState({ showModalNewAchievements: false });
  }

  showEventCompletedModal = (title) => {
    this.setState({ showEventCompleted: true, eventTitle: title });
  }

  hideEventCompletedModal = () => {
    this.removeQuizResult();
    this.setState({ showEventCompleted: false });
  }

  removeQuizResult() {
    this.props.removeQuizResult();
  }

  getTabletScale(zoomScale) {
    if (zoomScale === 1) {
      return 1.3;
    } else {
      return 2;
    }
  }

  render() {
    const { loading, level, zoomScale, hasUserTree, modalVisible, showModalNewAchievements, showEventCompleted, eventTitle, achievements, showLabelLayer } = this.state;
    const { device: { dimensions: { width, height, orientation } }, quizResult } = this.props;

    if (quizResult && (!showModalNewAchievements && !showEventCompleted)) {
      this.validateResultQuiz();
    }

    const showAchievementsModal = !modalVisible && showModalNewAchievements;
    const showEventCompletedModal = !modalVisible && !showModalNewAchievements && showEventCompleted;


    const videoDimensions = {
      width: 1280,
      height: 720
    };

    const minZoom = {
      scale: isTablet ? this.getTabletScale(zoomScale) : 1,
    }

    const zoomContainerStyle = isTablet ? {
      position: 'absolute',
      height: zoomScale === 1 ? '87%' : '67%',
      top: 0,
      left: 0,
      right: 0,
    } : {
      flex: 1,
    };

    if (loading && !hasUserTree) { return <Preloader />; }
    return (
      <TreeContainer>
        <AnimatedNubes deviceWidth={width} deviceHeight={height} orientation={orientation} />
        {showAchievementsModal && <NewAchievementsModal achievements={achievements} onHideModal={this.hideModalNewAchievements} />}
        {showEventCompletedModal && <EventCompletedModal eventTitle={eventTitle} onHideModal={this.hideEventCompletedModal} />}
        {!modalVisible &&
          <View style={{
            transform: [{ scale: minZoom.scale }],
            ...zoomContainerStyle,
          }}>
            <Zoom
              flex={Platform.OS === 'android' ? 10 : 1}
              maxScale={zoomScale}
              onViewTransformed={this.onViewTransformed}
              onTransformGestureReleased={this.onTransformGestureReleased}
              ref={view => this.treeView = view}
            >
              <MacetaContainer><Maceta width={200} /></MacetaContainer>
              {level}
            </Zoom>
          </View>
        }
        {(!modalVisible && showLabelLayer) && <LabelsLayer />}
        {modalVisible && <Video
          videoDimensions={videoDimensions}
          source={vineta_1}
          dismiss={() => this.showVideo(false)}
          visible={modalVisible}
          width={width}
          onPlaybackStatusUpdate={this.onPlaybackStatusUpdate}
          showCloseIcon={false}
          skipButton
        />}
      </TreeContainer>
    );
  }
}

const Zoom = styled(ViewTransformer)`
  overflow: visible;
  position: relative;
  flex: ${props => props.flex};
`;

Tree.propTypes = {
  device: PropTypes.any,
  userTree: PropTypes.object,
};
