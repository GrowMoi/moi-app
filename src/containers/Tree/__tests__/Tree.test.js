import React from 'react';
import { shallow } from 'enzyme';
// import renderer from 'react-test-renderer';
import Tree from '../Tree';

describe('Testing Tree View', () => {
  it('renders without crashing and take a snapshot of the structure', () => {
    const wrapper = shallow(<Tree />);
    // const rendered = renderer.create(<Home />).toJSON();
    const render = wrapper;
    expect(render).toMatchSnapshot();
    expect(render).toBeTruthy();
  });
});
