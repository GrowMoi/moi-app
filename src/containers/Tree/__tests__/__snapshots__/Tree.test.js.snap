// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Tree View renders without crashing and take a snapshot of the structure 1`] = `
<Styled(View)>
  <Navbar />
  <TouchableOpacity
    activeOpacity={0.2}
    focusedOpacity={0.7}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      color="white"
      disabled={false}
      ellipsizeMode="tail"
      onPress={[Function]}
    >
      Press Me!
    </Text>
  </TouchableOpacity>
</Styled(View)>
`;
